### PR TITLE
DRAFT: Background Tasks Scheduler

### PR DESCRIPTION
**Pull request by aarani**

### CODE REVIEW

**Comment by knocte** [comment]
Good work so far!

File: Xamarin.Essentials/Background/Background.android.service.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+namespace Xamarin.Essentials.Background
+{
+    [Service]
+    public class BackgroundService : Service
+    {
+        static bool isRunning;
+
+        public override IBinder OnBind(Intent intent)
+        {
+            return null;
```
Comment (knocte) [inline Xamarin.Essentials/Background/Background.android.service.cs:17]: null? is this correct?

File: Xamarin.Essentials/Background/Background.ios.service.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            if (isRunning)
+                return;
+
+            taskId = UIApplication.SharedApplication.BeginBackgroundTask("BackgroundTask", Stop);
+            Background.StartJobs();
+
+            isRunning = true;
+        }
+
+        public static void Stop()
```
Comment (knocte) [inline Xamarin.Essentials/Background/Background.ios.service.cs:27]: seems this method is not needed?

File: Xamarin.Essentials/Background/Background.android.service.cs (hunk 1 above)
Comment (aarani) [reply Xamarin.Essentials/Background/Background.android.service.cs:17]: These codes are copied from https://github.com/nblockchain/XamarinFormsBackground assessing accuracy of this code probably needs more testing

File: Xamarin.Essentials/Background/Background.ios.service.cs (hunk 2 above)
Comment (aarani) [reply Xamarin.Essentials/Background/Background.ios.service.cs:27]: Yes it's not implemented yet but I suppose we could use it to add features soon so I didn't went out of my way to get rid of it

Comment (knocte) [reply Xamarin.Essentials/Background/Background.ios.service.cs:27]: then let's remove the method, we'll add it when we need it

Comment (knocte) [reply Xamarin.Essentials/Background/Background.ios.service.cs:27]: but if Stop() is not called anywhere, even if you add something to it, it's unused, I'd say remove it until we need it

File: Xamarin.Essentials/Background/Background.tizen.cs [hunk 3]
```diff
@@ -8,7 +8,7 @@ public partial class Background
     {
         public static void StartBackgroundService()
         {
-            Background.StartJobs();
+            throw new NotImplementedException();
```
Comment (knocte) [inline Xamarin.Essentials/Background/Background.tizen.cs:12]: @aarani you forgot to add the TODO with the URL

**Comment by knocte** [comment]
but if Stop() is not called anywhere, even if you add something to it, it's unused, I'd say remove it until we need it

File: Xamarin.Essentials/Background/Background.ios.service.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using UIKit;
+
+namespace Xamarin.Essentials.Background
+{
+    public class BackgroundService
```
Comment (knocte) [inline Xamarin.Essentials/Background/Background.ios.service.cs]: class should be marked as static

**Comment by aarani** [comment]
> but if Stop() is not called anywhere, even if you add something to it, it's unused, I'd say remove it until we need it

`taskId = UIApplication.SharedApplication.BeginBackgroundTask("BackgroundTask", Stop);`

It's gonna get called by the OS when a task takes too long to process
I think we need a cancellation token here too to cancel whatever hapenning in the background if Stop() happend

**Comment by knocte** [comment]
> It's gonna get called by the OS when a task takes too long to process

I don't understand, and what happens if you remove the Stop method? It's not required at compilation time. Does Xamarin.iOS call the Stop method via reflection? I doubt it.

**Comment by aarani** [comment]
> > It's gonna get called by the OS when a task takes too long to process
> 
> I don't understand, and what happens if you remove the Stop method? It's not required at compilation time. Does Xamarin.iOS call the Stop method via reflection? I doubt it.

It is required at compilation time! what code are you looking at? Background.ios.service.cs#L21 Stop() is passed as an Action

**Comment by knocte** [comment]
> It is required at compilation time! what code are you looking at? Background.ios.service.cs#L21 Stop() is passed as an Action

Ah shit my bad, let's change `Stop` to `this.Stop` to make this more clear.

**Comment by knocte** [comment]
Nevermind, it's static!

**Pull request merged**